### ISSUE TITLE
B006 doesn’t detect mutable defaults in tuples, generators, and assignment expressions

### ISSUE THREAD

**Issue by dscorbett** [astral-sh/ruff#20004]
### Summary

[mutable-argument-default (B006)](https://docs.astral.sh/ruff/rules/mutable-argument-default/) doesn’t detect certain guaranteed-mutable expressions which should be simple to support: tuples containing mutable expressions, generator expressions, and assignment expressions containing mutable expressions. [Example](https://play.ruff.rs/280e1f9b-99c8-4c79-8d27-0e9bbff72f34):
```console
$ cat >b006.py <<'# EOF'
def f(x=([],), y=(y for y in "y"), z=(z := [])):
    pass
# EOF

$ ruff --isolated check b006.py --select B006
All checks passed!
```

### Version

ruff 0.12.9 (ef422460d 2025-08-14)

**Comment by IDrokin117** [comment]
I am going to work on this

**Comment by IDrokin117** [comment]
Supporting assignment expressions appears to be challenging for non-obvious reasons. Consider the function
```python
def f(z=(q := [])):
    print(z)
```
[AST node `NamedExpr`](https://play.ruff.rs/da6feb34-375e-4441-96d8-f6df66e3420c) range captures only the inner value `q := []` instead of `(q := [])`. That means, if we try to fix the code according to the default procedure, it will introduce a syntax error, because parentheses are ignored
```python
def f(z=(None)): # Parentheses are left
    if z is None:
        z = q := [] # Parentheses are missing
    print(z)
```
While for other cases it is ok 
```python
# Before fix
def f(z=([])):
    print(z)
```
```python
# After fix
def f(z=(None)): # Parentheses left
    if z is None:
        z = []
    print(z)
```
but for `NamedExpr` it is not OK, because parentheses are mandatory. I guess parentheses must be part of the range, e.g. `NamedExpr("(q := [])")` or somehow signal about them.

All in all, I propose the following options:
1. Do not add "assignment expressions containing mutable expressions" to this rule.
2. Modify the `NamedExpr` range to include parentheses or introduce a metadata flag.. (not sure about this)
3. Add parentheses to the fix when the default value is a `NamedExpr`, while preserving the original parentheses. Ideally, the original parentheses should be removed since they are semantically bound to the `NamedExpr`, but given the implementation complexity, I favor leaving them unchanged.
```python
def f(z=(None)):
    if z is None:
        z = (q := [])
    print(z)
```

@ntBre What you think?

**Comment by dscorbett** [comment]
Fixing the assignment expression is tricky, but B006 should still be able to detect it without a fix.

**Comment by ntBre** [comment]
Could we use [`parenthesized_range`](https://github.com/astral-sh/ruff/blob/692be72f5aff65bfea504504c54d82112657c98a/crates/ruff_python_ast/src/parenthesize.rs#L60) here? That seems like it could handle both removing and adding parentheses, but I haven't looked at the rule implementation.

I'm also happy with detecting the issue and not offering a fix if it's too complicated, as @dscorbett suggested.
